modules/src/core/ics26_routing/context.rs

```diff
@@ -17,4 +34,188 @@ pub trait Ics26Context:
     + PortReader
     + Ics20Context
 {
+    type Router: Router;
+
+    fn router(&self) -> &Self::Router;
+
+    fn router_mut(&mut self) -> &mut Self::Router;
+}
+
+#[derive(Debug, PartialEq)]
+pub struct InvalidModuleId;
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ModuleId(String);
+
+impl ModuleId {
+    pub fn new(s: Cow<'_, str>) -> Result<Self, InvalidModuleId> {
+        if !s.trim().is_empty() && s.chars().all(char::is_alphanumeric) {
+            Ok(Self(s.into_owned()))
+        } else {
+            Err(InvalidModuleId)
+        }
+    }
+}
+
+impl fmt::Display for ModuleId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl FromStr for ModuleId {
+    type Err = InvalidModuleId;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::new(Cow::Borrowed(s))
+    }
+}
+
+impl Borrow<str> for ModuleId {
+    fn borrow(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+pub trait Acknowledgement: AsRef<[u8]> {
```

Comment:
Similarly, would it be worth having this in the docstring?

Reply:
Just added a doc comment -> https://github.com/informalsystems/ibc-rs/pull/1838/commits/df4e96872490d27be550a43caaef126821d4161f
IIUC, the problem with object safety is that we'd like to return `dyn Acknowledgement` from `Module::on_recv_packet()`, and the reason we can't use a generic type there is because `Module` needs to be trait object safe so we can have a collection of modules in the router. `From` requires `Self: Sized`, so it is incompatible with trait objects.